### PR TITLE
Document the ref keyword

### PR DESCRIPTION
**Pull request by poliorcetics**
Partial fix for #34601.

This documents the `ref` keyword with two examples, one failing to compile because the `ref` keyword is missing, and the same example fixed with the keyword inserted in the correct place.

It also explains (very *very* rapidly) the differences between `&` and `ref`.

I put a link to the best place I could find in the Reference but there may be something better that I didn't find.

### CODE REVIEW

File: src/libstd/keyword_docs.rs [hunk 1]
```diff
@@ -991,9 +991,49 @@ mod pub_keyword {}
 //
 /// Bind by reference during pattern matching.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+/// `ref` annotates pattern bindings to make them borrow rather than move.
+/// It is **not** a part of the pattern as far as matching is concerned.
```
Comment (jyn514) [inline src/libstd/keyword_docs.rs:995]: I think I understand what you're trying to say - the same values are matched with or without `ref` - but it's not very clear from the wording. Maybe you can say "this does not affect _whether_ a value is matched, only _how_ it is matched"?

File: src/libstd/keyword_docs.rs [hunk 2]
```diff
@@ -991,9 +991,49 @@ mod pub_keyword {}
...
+///     Some(n) => println!("Hello, {}", n),
+///     _ => println!("Hello, world"),
+/// }
+/// // ... and now unavailable.
+/// println!("Hello again, {}", maybe_name.unwrap_or("world".into()));
+/// ```
 ///
-/// [not yet complete]: https://github.com/rust-lang/rust/issues/34601
+/// Using the `ref` keyword, the value is only borrowed, not moved, making
+/// available for use after the [`match`] statement:
```
Comment (jyn514) [inline src/libstd/keyword_docs.rs:1012]: I didn't know this! Great explanation :)

File: src/libstd/keyword_docs.rs (hunk 1 above)
Comment (poliorcetics) [reply src/libstd/keyword_docs.rs]: Done 👍

File: src/libstd/keyword_docs.rs (hunk 2 above)
Comment (poliorcetics) [reply src/libstd/keyword_docs.rs:1013]: Glad to hear it 😄

File: src/libstd/keyword_docs.rs [hunk 3]
```diff
@@ -991,9 +991,50 @@ mod pub_keyword {}
...
+/// be a problem, when you don't really need the value to be moved and owned:
+///
+/// ```compile_fail,E0382
+/// let maybe_name = Some(String::from("Alice"));
+/// // The variable 'maybe_name' is consumed here ...
+/// match maybe_name {
+///     Some(n) => println!("Hello, {}", n),
+///     _ => println!("Hello, world"),
+/// }
+/// // ... and now unavailable.
```
Comment (jyn514) [inline src/libstd/keyword_docs.rs:1008]:
```suggestion
/// // ... and is now unavailable.
```

File: src/libstd/keyword_docs.rs [hunk 4]
```diff
@@ -991,9 +991,50 @@ mod pub_keyword {}
...
+/// println!("Hello again, {}", maybe_name.unwrap_or("world".into()));
+/// ```
 ///
-/// [not yet complete]: https://github.com/rust-lang/rust/issues/34601
+/// Using the `ref` keyword, the value is only borrowed, not moved, making
+/// available for use after the [`match`] statement:
```
Comment (jyn514) [inline src/libstd/keyword_docs.rs:1013]:
```suggestion
/// it available for use after the [`match`] statement:
```

File: src/libstd/keyword_docs.rs [hunk 5]
```diff
@@ -991,9 +991,50 @@ mod pub_keyword {}
...
+///
+/// ```
+/// let maybe_name = Some(String::from("Alice"));
+/// // Using `ref`, the value is borrowed, not moved ...
+/// match maybe_name {
+///     Some(ref n) => println!("Hello, {}", n),
+///     _ => println!("Hello, world"),
+/// }
+/// // ... and it's available here !
```
Comment (jyn514) [inline src/libstd/keyword_docs.rs:1022]:
```suggestion
/// // ... so it's available here!
```

**Review by jyn514 — APPROVED** [review]
r=me once nits are fixed

File: src/libstd/keyword_docs.rs (hunk 5 above)
Comment (poliorcetics) [reply src/libstd/keyword_docs.rs]: Fixed all of them. 😄

**Pull request merged**